src/main/java/com/rabbitmq/client/impl/nio/SocketChannelFrameHandlerState.java

```diff
@@ -176,11 +180,14 @@ void endWriteSequence() {
 
     void prepareForReadSequence() throws IOException {
         if(ssl) {
-            cipherIn.clear();
-            plainIn.clear();
+            if (!isUnderflowHandlingEnabled.get()) {
+                cipherIn.clear();
+                cipherIn.flip();
+            }
 
-            cipherIn.flip();
+            plainIn.clear();
             plainIn.flip();
```

Comment:
Can you elaborate on the need of these operations?

Reply:
`cipherIn` should aggregate chunks of bytes on each read. So we should not clear it before each chunk reading.

Reply:
Also in my implementation `continueReading` returns false when there is no data for next chunk. So there is no infinite loop around `while(continueReading())`. NIO thread can check selector state before trying to get bytes.